Bug753307 - Custom Report be selectable as default Report for Printing

Bob-IT opened the pull request:
This commit changes the preference in Business->'Report for Printing' to be saved as a book property and allow the selection of any Invoice Report to be used as the default. This still needs work...

At the moment it works by returning all reports and I have successfully selected a new report scheme file, saved configuration and the existing 4 invoice reports. If the saved report can not be found, I have added an entry to the combo list with the missing guid and present a warning image with tool tip.

I am not convinced I have added to the correct file, 'dialog-utils.*', I think it probably should be in 'business-gnome-utils.*' but that is in the wrong directory for 'dialog-preference' to access it. I have not tried moving 'business-gnome-utils.*' to the 'gnome-utils' folder but that may be an option.

@christopherlam  Can you have a look at the required scheme changes to filter the list, at the moment I was able to work out how to return all reports but this should be limited to just invoice ones. Also for the function 'gnc_invoice_window_print_invoice', a test needs to be made for valid returned guid, maybe you can add some scheme code to return true/false to be combined with the test for 'reportname'.

Any further changes I make will be after the required scheme changes to make it useful.

As an after thought, I did wonder if this sort of change could be added for other similar use reports but the only one that may make sense is the 'Account Report'.

christopherlam commented on gnucash/gnome-utils/dialog-utils.c:
```diff
@@ -876,3 +879,187 @@ gnc_get_negative_color (void)
+        gtk_tree_model_get (model, &iter, COL_INV_NAME, &name,
+                                          COL_INV_GUID, &guid, -1);
+
+        if (g_strcmp0 (name, _("Missing Invoice Report")) == 0)
+            gtk_widget_show (user_data);
+        else
+            gtk_widget_hide (user_data);
```
you don't use `gtk_widget_set_visible`? I'm usually one for brevity...

Bob-IT replied:
That maybe factored out on later revisions, on first passes I tend to see what works.

christopherlam reviewed (commented):
Here's a patch; note it modifies existing reports to add an `'invoice` tag. This may not be desirable. TODO: ensure saved-reports also receive this `'invoice` tag. To be discussed. Another hackish approach is to generate options, and query them to look for "Invoice Number" option.

```patch
diff --git a/gnucash/report/report-core.scm b/gnucash/report/report-core.scm
index 7368bd397..8f2100c5c 100644
--- a/gnucash/report/report-core.scm
+++ b/gnucash/report/report-core.scm
@@ -170,7 +170,7 @@
   (make-new-record-template version name report-guid parent-type options-generator
                             options-cleanup-cb options-changed-cb
                             renderer in-menu? menu-path menu-name
-                            menu-tip export-types export-thunk)
+                            menu-tip hook export-types export-thunk)
   report-template?
   (version report-template-version)
   (report-guid report-template-report-guid report-template-set-report-guid!)
@@ -184,11 +184,12 @@
   (menu-path report-template-menu-path)
   (menu-name report-template-menu-name)
   (menu-tip report-template-menu-tip)
+  (hook report-template-hook)
   (export-types report-template-export-types)
   (export-thunk report-template-export-thunk))
 
 (define (make-report-template)
-  (make-new-record-template #f #f #f #f #f #f #f #f #t #f #f #f #f #f))
+  (make-new-record-template #f #f #f #f #f #f #f #f #t #f #f #f #f #f #f))
 (define gnc:report-template-version report-template-version)
 (define gnc:report-template-report-guid report-template-report-guid)
 (define gnc:report-template-set-report-guid! report-template-set-report-guid!)
@@ -204,6 +205,7 @@
 (define gnc:report-template-menu-path report-template-menu-path)
 (define gnc:report-template-menu-name report-template-menu-name)
 (define gnc:report-template-menu-tip report-template-menu-tip)
+(define gnc:report-template-hook report-template-hook)
 (define gnc:report-template-export-types report-template-export-types)
 (define gnc:report-template-export-thunk report-template-export-thunk)
 
@@ -480,8 +482,9 @@ not found.")))
 
 ;; return a list of the invoice report template guids.
 (define (gnc:custom-report-invoice-template-guids)
-  (map car (hash-map->list cons *gnc:_report-templates_*)))
-;;  (map car (gnc:custom-report-invoice-templates-list)))
+  (hash-fold
+   (lambda (k v p) (if (eq? (gnc:report-template-hook v) 'invoice) (cons k p) p))
+   '() *gnc:_report-templates_*))
 
 (define (gnc:find-report-template guid)
   (hash-ref *gnc:_report-templates_* guid))
diff --git a/gnucash/report/reports/standard/invoice.scm b/gnucash/report/reports/standard/invoice.scm
index 0b0e46557..ebd401b08 100644
--- a/gnucash/report/reports/standard/invoice.scm
+++ b/gnucash/report/reports/standard/invoice.scm
@@ -892,6 +892,7 @@ for styling the invoice. Please see the exported report for the CSS class names.
  'menu-path (list gnc:menuname-business-reports)
  'options-generator (lambda () (options-generator 'invoice))
  'renderer reg-renderer
+ 'hook 'invoice
  'in-menu? #t)
 
 (gnc:define-report
@@ -901,6 +902,7 @@ for styling the invoice. Please see the exported report for the CSS class names.
  'menu-path (list gnc:menuname-business-reports)
  'options-generator (lambda () (options-generator 'easy-invoice))
  'renderer reg-renderer
+ 'hook 'invoice
  'in-menu? #t)
 
 (gnc:define-report
@@ -910,5 +912,6 @@ for styling the invoice. Please see the exported report for the CSS class names.
  'menu-path (list gnc:menuname-business-reports)
  'options-generator (lambda () (options-generator 'fancy-invoice))
  'renderer reg-renderer
+ 'hook 'invoice
  'in-menu? #t)
 
diff --git a/gnucash/report/reports/standard/taxinvoice.scm b/gnucash/report/reports/standard/taxinvoice.scm
index 7992d871e..fdcd81ed1 100644
--- a/gnucash/report/reports/standard/taxinvoice.scm
+++ b/gnucash/report/reports/standard/taxinvoice.scm
@@ -289,6 +289,7 @@
   'menu-name (N_ "Tax Invoice")
   'menu-tip (N_ "Display a customer invoice with tax columns (using eguile template)")
   'menu-path (list gnc:menuname-business-reports)
+  'hook 'invoice
   'options-generator options-generator
   'renderer report-renderer)
 
@@ -316,5 +317,6 @@
   'menu-name (N_ "Australian Tax Invoice")
   'menu-tip (N_ "Display an Australian customer invoice with tax columns (using eguile template)")
   'menu-path (list gnc:menuname-business-reports)
+  'hook 'invoice
   'options-generator au-tax-options-generator
   'renderer report-renderer)
```

christopherlam commented:
Here's my TODO done: searches built-in *and* saved-reports for invoice types. It completes the C code. A saved invoice report is selectable as the default for invoice printing. If the saved-report is deleted, it'll degrades to the default "Printable Invoice" report.

```scheme
;; return a list of the invoice report template guids.
(define (gnc:custom-report-invoice-template-guids)
  (hash-fold
   (lambda (k v p)
     (if (eq? (gnc:report-template-hook v) 'invoice)
         (cons k p)
         (let ((parent-type (gnc:report-template-parent-type v)))
           (if parent-type
               (if (eq? (gnc:report-template-hook
                         (hash-ref *gnc:_report-templates_* parent-type)) 'invoice)
                   (cons k p)
                   p)
               p))))
   '() *gnc:_report-templates_*))
```

christopherlam commented:
I think you wanted a function to verify a report guid is a valid invoice report? The `gnc:report-is-invoice-report?` can be called from `gnc_invoice_window_print_invoice`:

```patch
Unstaged changes (1)
modified   gnucash/report/report-core.scm
@@ -481,16 +481,19 @@ not found.")))
 (define (gnc:custom-report-template-guids)
   (map car (gnc:custom-report-templates-list)))
 
+(export gnc:report-is-invoice-report?)
+(define (gnc:report-is-invoice-report? guid)
+  (let ((tmpl (gnc:find-report-template guid)))
+    (cond
+     ((not tmpl) #f)
+     ((eq? (gnc:report-template-hook tmpl) 'invoice) #t)
+     (else (and-let* ((type (gnc:report-template-parent-type tmpl))
+                      (template (gnc:find-report-template type)))
+             (eq? (gnc:report-template-hook template) 'invoice))))))
+
 ;; return a list of the invoice report template guids.
 (define (gnc:custom-report-invoice-template-guids)
   (hash-fold
-   (lambda (k v p)
-     (if (or (eq? (gnc:report-template-hook v) 'invoice)
-             (and-let* ((type (gnc:report-template-parent-type v))
-                        (template (gnc:find-report-template type)))
-               (eq? (gnc:report-template-hook template) 'invoice)))
-         (cons k p)
-         p))
+   (lambda (k v p) (if (gnc:report-is-invoice-report? k) (cons k p) p))
    '() *gnc:_report-templates_*))
 
 (define (gnc:find-report-template guid)
```

fellen commented on gnucash/gnome-utils/dialog-utils.c:
```diff
@@ -876,3 +879,187 @@ gnc_get_negative_color (void)
+
+        gtk_list_store_prepend (GTK_LIST_STORE(model), &iter);
+        gtk_list_store_set (GTK_LIST_STORE(model), &iter,
+                            COL_INV_NAME, _("Missing Invoice Report"),
```
Can you add a translator comment where it appears, resp. why it is capitalized?

Bob-IT replied:
I have decided to save the report name also so this has changed.

Bob-IT commented:
@christopherlam Thanks for the scheme changes, I have added them and changed my original push to save the default report name as well as the guid to make the message more meaningful when it is missing. It all looks OK and I think works well but maybe you could give it a spin.

I think I can move my 'dialog-utils' changes to 'business-gnome-utils' in the gnome directory, just need to prove the callbacks will still work from the preference dialog.

@gjanssens If you have time to comment, would be appreciated.

christopherlam commented:
My last UI/usability comment is that a book specific preference belongs on the book properties dialog rather than the preferences dialog. Otherwise, wfm!

Bob-IT commented:
> My last UI/usability comment is that a book specific preference belongs on the book properties dialog rather than the preferences dialog.

I did wonder about that, if set in stone that is a totally different ball game. It would then need something similar to 'gnc:make-budget-option' which is more scheme, not sure how that would work out.

jralls commented:
> It would then need something similar to 'gnc:make-budget-option' which is more scheme, not sure how that would work out.

Not necessarily in scheme, just (for now) callable from scheme, meaning that the new function would have to be swigged. When `c++ options` is in place there will be no scheme in the book options execution path.

Is this really a small enough feature to go into maint?

Bob-IT commented:
> Is this really a small enough feature to go into maint?

It started of small not really knowing what it would involve. I am in no hurry to push this so I am quite happy to pause this and wait for your c++ options changes and resume this on master.

jralls commented:
OK. On the topic of `c++options` could you have a look over the [code](https://github.com/jralls/gnucash/tree/c+%2Boptions), particularly `dialog-options.cpp`?

fellen commented on gnucash/gnome-utils/dialog-utils.c:
```diff
@@ -876,3 +879,192 @@ gnc_get_negative_color (void)
+    {
+        GtkTreeIter iter;
+        /* Translators: %s is the default invoice report name. */
+        gchar *tool_tip = g_strdup_printf (_("'%s' is missing"),
```
Full stop?

HIG and Weblate like grammatical correct sentences.

Bob-IT replied:
Thanks, missed that, I also have a post-it note to remind me!
Will change.

Bob-IT commented:
@jralls I tried to merge your c++ options branch to master on Sunday which highlighted some non initialized and missing declarations. I see you have already fixed what I had noticed but was wondering how much is working for you. Just tried 'File->Properties' and resulted in an empty dialog and trying to run a report ends in a segmentation fault.

jralls commented:
@Bob-IT, the current status is that c++options finally passes all of the report unit tests as of yesterday (except for one weird failure on github's Ubuntu 18.04 VM that passes on my 18.04 VM) so I've just begun testing the GUI components. It's definitely not working yet.

I want you to look at the GUI code changes, particularly the rewrite of dialog-options.cpp, to see if the design makes sense to you.

There's one detail question as well: In several cases we wrap the UI element in a GtkEventBox with the comment that it's to enable the per-choice tooltip. Is that still necessary now that we've removed the per-choice tooltips from multichoice options?

Bob-IT commented:
OK, will try and have a look tomorrow.

Bob-IT commented:
@jralls Regarding the GtkEventBox's, there are three places where it is added and I tried removing all of them and all tool tips still worked so I think they can safely be removed. Will have a look at dialog-options.cpp to see if I can get an inkling of the changes.

Bob-IT commented:
@jralls I have had a look at dialog-options.cpp and I think I have an idea of what is going on and I think the design/layout is good and seems simpler.
I did notice a couple of changes that I think are needed for create_multichoice_widget and RelativeDateEntry. Which are, the list store should have one string, gtk_list_store_set should end in -1); and I think the cell renderer setup/layout is missing.
Will wait for it to be working to see if I can put the pieces together that are underpinning these changes.

jralls commented:
@Bob-IT I made some good progress today and fixed it enough that it now loads the balance sheet report (and maybe others) and then displays the options dialog without crashing. Out of time for today so didn't test the options dialog controls.

jralls commented on gnucash/gnome-utils/dialog-utils.c:
```diff
@@ -876,3 +879,192 @@ gnc_get_negative_color (void)
+gnc_get_default_invoice_print_report (void)
+{
+    QofBook *book = gnc_get_current_book ();
+    int value = gnc_prefs_get_int (GNC_PREFS_GROUP_INVOICE, GNC_PREF_INV_PRINT_RPT);
+    const gchar *default_guid = gncInvoiceGetDefaultReport (book);
+
+    if (default_guid == NULL)
+    {
+        if (value >= 0 && value < 4)
```
Dunno about citizenship, but you're  right that if a variable is used only in a scope it should be declared in that scope.

gjanssens reviewed (changes requested):
After months I have finally found the time to look at this (and partly prompted because of PR #1247 ).

I think this is generally a very useful step forward. I've added a few remarks.

Considering this PR together with #1247 I am wondering if it would be worth abstracting the design into "a specialized combobox widget that lists reports and saved reports" (say `GncReportCombo`) so it could be reused in both use cases.

* This PR is about setting a default (saved) report to use for invoice printing. So it could use that more specialized combo box and build functionality around it to get or set a preference.
* PR #1247 on the other hand is about occasionally using **another** (saved) report to print a certain invoice (without making a detour via an invoice report window). That feature could also benefit from a more specialized combo box and use it to get a one-time value for a one-off (saved) report to use.

This doesn't have to be done as part of this PR necessarily though. I'm fine with the feature as implemented right now already (barring my few comments inline) and I think many users would be quite happy with this improvement.

Working towards a specialized combo box can be done as a next step. The data to store for this particular PR wouldn't change, only the code organisation would. (I'd be much more reluctant if a future adaptation would involve data changes).

gjanssens commented:
> > My last UI/usability comment is that a book specific preference belongs on the book properties dialog rather than the preferences dialog.
> 
> I did wonder about that, if set in stone that is a totally different ball game. It would then need something similar to 'gnc:make-budget-option' which is more scheme, not sure how that would work out.

That remark makes me regret I refused your prototype to redo the book preferences dialog in C. Back then I had high theories about how it's better to keep option functionality and gui together. Now I notice this just makes the barrier to move options much higher. It's true that it's quite complex to add a new book option in the book option gui at the moment.

So I'd be tempted to reconsider your glade based book options dialog :confused:

Bob-IT commented on gnucash/gnome-utils/dialog-utils.c:
```diff
@@ -876,3 +879,192 @@ gnc_get_negative_color (void)
+    QofBook *book = gnc_get_current_book ();
+    const gchar *default_guid = gnc_get_default_invoice_print_report ();
+    const gchar *default_name = gncInvoiceGetDefaultReportName (book);
+    gboolean have_default = FALSE;
```
I missed that but have changed this.

Bob-IT commented on gnucash/gnome-utils/dialog-utils.c:
```diff
@@ -876,3 +879,192 @@ gnc_get_negative_color (void)
+    "67112f318bef4fc496bdc27d106bbda4", // "Easy Invoice"
+    "3ce293441e894423a2425d7a22dd1ac6", // "Fancy Invoice"
+    NULL
+};
```
I have added enums and added some #define's as I could not get a const array of const variables to work.

Bob-IT commented:
This change has been moved to master under PR #1333 so closing.
